File: languages/clojure/destructuring_and_pattern_matching.clj
(def a-list [1,2,3,4,5])

(println a-list)

; Destructuring a list
; Tuple can be represented using a List
(println (let [[first second] a-list]
     [first, second]))
     
(let [
      [first second & rest] a-list
     ]
     (println [first, second, rest]))


; Destructuring a map (Associative Destructuring)
(def a-name {:first "Dhaval" :last "Dalal" :salutation "Mr."})

(println a-name)

(let [
      {fname :first lname :last salutation :salutation} a-name
     ]
  (println fname lname))

(require '[clojure.string :as string])

(defn capitalize [{fname :first lname :last}] (str (string/upper-case fname) " " (string/upper-case lname)))

(println (capitalize a-name))

; Idiomatic Clojure does not have pattern matching like that in Scala, 
; but we can use multi-methods to achieve that.

File: languages/scala/destructuring_and_pattern_matching.scala
// In FP, Destructuring is about extracting data from the innards of any structure.

// Destructuring list
val list = 1::2::3::4::5::6::Nil
// Destructuring gives us a short-hand way of naming parts of the data-structure.
val (first::second::rest) = list
println(first)
println(second)
println(rest)

// Destructuring nTuples
val (x, y) = (20, 30)
println(x)
println(y)

val (p, q, r) = (20, 30, 40)
println(p)
println(q)
println(r)

case class Name(fName: String, lName: String, salutation: String)

// Pattern Matching
def capitalize(name: Name) = name match {
  case Name(firstName, lastName, _) => firstName.toUpperCase + " " + lastName.toUpperCase
}

println(capitalize(Name("Dhaval", "Dalal", "Mr.")))

type Point = (Int, Int)
// Pattern Matching
def midpoint(p1: Point, p2: Point): Point = {
  def mid(a1: Int, a2: Int) = (a1 + a2) / 2
  (p1, p2) match {
    case (p1, (0, 0)) => p1
    case ((0, 0), p2) => p2
    case (p1, p2) if p1 == p2 => p1
    case ((x1, y1), (x2, y2)) => (mid(x1, x2), mid(y1, y2))
  }
}

println(midpoint((2,3), (4,5)))
println(midpoint((0,0), (4,5)))
println(midpoint((2,3), (0,0)))
println(midpoint((2,3), (2,3)))